optimum/gptq/quantizer.py

```diff
@@ -0,0 +1,629 @@
+                )
+        if self.dataset is None:
+            raise ValueError("You need to pass `dataset` in order to quantize your model")
+        elif isinstance(self.dataset, str):
+            dataset = get_dataset(self.dataset, tokenizer, seqlen=self.model_seqlen, split="train")
+        elif isinstance(self.dataset, list):
+            dataset = [tokenizer(data, return_tensors="pt") for data in self.dataset]
+        else:
+            raise ValueError("You need to pass a list of string or a string for `dataset`")
+
+        dataset = prepare_dataset(dataset, pad_token_id=self.pad_token_id, batch_size=self.batch_size)
+
+        # Step 2: get the input of the 1st block
+        # To do that, we need to put the modules preceding the first block on the same device as the first bloc.
+        # Then we run the model and it will stop at the first bloc as we added a prehook that raise an Exception after storing the inputs.
+
+        layer_inputs = []
+        layer_outputs = []
+        layer_input_kwargs = []
+
+        if self.block_name_to_quantize is None:
+            self.block_name_to_quantize = get_block_name_with_pattern(model)
+
+        if self.module_name_preceding_first_block is None:
+            self.module_name_preceding_first_block = get_preceding_modules(model, self.block_name_to_quantize)
+
+        blocks = recurse_getattr(model, self.block_name_to_quantize)
+
+        if not has_device_map:
+            # put modules from module_name_preceding_first_block on cuda
+            for module_name in self.module_name_preceding_first_block:
+                module = recurse_getattr(model, module_name)
+                if module is None:
+                    raise ValueError(f"Module {module_name} was not found in model")
+                module = module.to(0)
+            blocks[0] = blocks[0].to(0)
+
+        def store_input_hook(_, input, *args):
+            kwargs = args[0]
+            input = input[0]
+            if input is None:
+                if "hidden_states" in kwargs:
+                    input = kwargs["hidden_states"]
+                else:
+                    raise ValueError("No input value found in the foward pass")
+            layer_inputs.append(input)
+            other_kwargs = {}
+            for k, v in kwargs.items():  # make sure other arguments also be captured
+                if k not in ["hidden_states"]:
+                    other_kwargs[k] = v
+            layer_input_kwargs.append(other_kwargs)
+            raise ValueError
+
+        handle = blocks[0].register_forward_pre_hook(store_input_hook, with_kwargs=True)
+        for data in dataset:
+            for k, v in data.items():
+                # put the data on gpu, we won't put them back to cpu
+                data[k] = v.to(0)
+            try:
+                model(**data)
+            except ValueError:
+                pass
+
+        handle.remove()
+        if not has_device_map:
+            blocks[0].to(device)
```

Comment:
Here `device` may be cpu, is it correct?

Reply:
yes !